R/RLearner_regr_gstat.R

```diff
@@ -0,0 +1,164 @@
+      makeUntypedLearnerParam(id = "model"), # FIXME what should be the type ?
+      makeUntypedLearnerParam(id = "range", default = NA),
+      makeNumericLearnerParam(id = "kappa", default = 0.5),
+      makeUntypedLearnerParam(id = "nugget"),
+      makeUntypedLearnerParam(id = "add.to"),
+      makeUntypedLearnerParam(id = "covtable"),
+      makeNumericLearnerParam(id = "Err", default = 0)
+    ),
+    par.vals = list(locations = ~x+y, nmax = 0, nmin = 0, omax = 0, maxdist = Inf,
+      dummy = FALSE, fill.all = FALSE, fill.cross = TRUE, variance = "identity", degree = 0, vdist = FALSE,
+      alpha = 0, beta.variogram = 0, cressie = FALSE, dX = 0, boundaries = 0, cloud = FALSE,
+      trend.beta = NULL, debug.level = 1, cross = TRUE, map = FALSE, projected = TRUE, lambda = 1.0,
+      verbose = FALSE, covariogram = FALSE, PR = FALSE, pseudo = -1,
+      fit.sills = TRUE, fit.ranges = TRUE, fit.method = 7, warn.if.neg = FALSE, fit.kappa = FALSE,
+      psill = NA, range = NA, kappa = 0.5, Err = 0),
+    properties = c("numerics", "factors" , "se", "weights", "missings"),
+    name = "Multivariable Geostatistical Prediction And Simulation",
+    short.name = "gstat",
+    note = "The default prediction (when passing no arguments) is IDW (inverse distance weighted).
+    The learner handles gstat variogram autofitting functionnality presented in this post https://www.r-spatial.org/r/2016/02/14/gstat-variogram-fitting.html.\n
+    Manual fitting example : lrn.man = makeLearner(cl = 'regr.gstat', id = 'manual', psill = 1, model = 'Sph', range = 900, nugget = 2, locations = ~x+y).\n
+    Automatic fitting example : lrn.auto = makeLearner(cl = 'regr.gstat', id = 'auto', psill = c('Sph','Exp','Gau', 'Mat'), locations = ~x+y)"
+  )
+}
+
+#' @export
+# https://stackoverflow.com/questions/19075331/passing-a-function-argument-to-other-arguments-which-are-functions-themselves
+# https://stackoverflow.com/questions/16774946/passing-along-ellipsis-arguments-to-two-different-functions
+trainLearner.regr.gstat = function(.learner, .task, .subset, .weights = NULL, ...) {
+  dots = list(...)
+  # https://stackoverflow.com/questions/11885207/get-all-parameters-as-list
+  variogram.names = names(formals(gstat::variogram)) # FIXME cannot retrieve the S3 method for formula arguments
+  variogram.names = c("object", "locations", "beta.variogram")
+  #variogram.names = replace(names(formals(gstat::variogram)), variogram.names == "beta", "beta.variogram")
+  fit.variogram.names = names(formals(gstat::fit.variogram))
+  gstat.names = replace(names(formals(gstat::gstat)), names(formals(gstat::gstat)) == "beta", "beta.gstat")
+
+  d = getTaskData(.task, .subset)
+  f = getTaskFormula(.task, explicit.features = TRUE)
+  # remove location vars as they are handled by gstat - https://stackoverflow.com/questions/40308944/removing-offset-terms-from-a-formula
+  f = update(f, .~.-y-x) # FIXME should be the params entered in locations arg
```

Comment:
It sounds like you want to check whether `x` and `y` are present and if not error or at least warn.

Reply:
I actually need to rewrite the formula as gstat "knows" (i.e. does not required coordinates in its the formula). My problem here is that `y` and `x` are hard coded. So my code assumes  that coordinates will always be referred as `x` and `y`. But it should also work for other names (as e.g. `lat`, `lon`) passed to the parameter with `id = location`. I need to find a way to dynamically retrieve the names of these parameters to update my formula.